File: lecture_1/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

bool check_sum(long number);
void print_card_logo(long number);

int main(void) {
    long number;

    do {
        number = get_long("Number: ");
    } while(!number || number < 0);

    int number_length = (int)(number ? log10(number) + 1 : 1);

    if ((number_length == 13 || number_length == 15 || number_length == 16) && check_sum(number)) {
        print_card_logo(number);
    } else {
        printf("INVALID\n");
    }
}

bool check_sum(long number) {
    int sum = 0;

    for (int i = 0; number !=0; i++, number /= 10) {
        if (i % 2 == 0){
            sum += number % 10;
        } else {
            int digit = 2 * (number % 10);
            sum += digit /10 + digit % 10;
        }
    }
    return (sum % 10) == 0;
}

void print_card_logo(long number){
    if  ((number >= 34e13 && number < 35e13) || (number >= 37e13 && number < 38e13)){
         printf("AMEX\n");
    }
    if  (number >= 51e14 && number < 56e14){
         printf("MASTERCARD\n");
    }
        if  ((number >= 4e12 && number < 5e12) || (number >= 4e15 && number < 5e15)){
         printf("VISA\n");
    }
}

File: lecture_6/credit/credit.py
import cs50
import math

def main ():
    while True:
        cardNumber = cs50.get_int("Enter Card Number (without spaces or dashes): ")

        if cardNumber and cardNumber > 0:
            break

    # get number length using math
    digits = int(math.log10(cardNumber))+1

    if (digits == 13 or digits == 15 or digits == 16) and checkSum(cardNumber):
        printCardLogo(cardNumber)
    else:
        print("INVALID")

# based on source code from geeksforgeeks.org/luhn-algorithm/
def checkSum(cardNumber):
    # get number length using converting to string
    cardNumberStr = str(cardNumber)
    cardNumberDigits = len(cardNumberStr)
    cardNumberSum = 0
    isSecond = False

    for i in range(cardNumberDigits - 1, -1, -1):
        d = ord(cardNumberStr[i]) - ord('0')

        if (isSecond == True):
            d = d * 2

        # We add two digits to handle
        # cases that make two digits after
        # doubling
        cardNumberSum += d // 10
        cardNumberSum += d % 10

        isSecond = not isSecond

    if (cardNumberSum % 10 == 0):
        return True
    else:
        return False

def printCardLogo(number):
    if (number >= 34e13 and number < 35e13) or (number >= 37e13 and number < 38e13):
        print("AMEX")
    elif  number >= 51e14 and number < 56e14:
        print("MASTERCARD")
    elif  (number >= 4e12 and number < 5e12) or (number >= 4e15 and number < 5e15):
        print("VISA")
    else:
        print("INVALID")

if __name__=="__main__":
    main()
